### ISSUE TITLE
Passing tags to tower-cli results in started job without tags

### ISSUE THREAD

**Issue by rvdh** [ansible/tower-cli#136]
Here I launch a new job based on template 272 with the tags default,npm,pm2,env:
```
$ ~/webhook/bin/tower-cli job launch --job-template 272 --tags default,npm,pm2,env -v -f json
*** DETAILS: Getting the record. **********************************************
GET https://192.168.4.14/api/v1/job_templates/272/
Params: {}

*** DETAILS: Asking for information necessary to start the job. ***************
GET https://192.168.4.14/api/v1/job_templates/272/launch/
Params: {}
```
Here you can see it's launching the job with the job_tags data:
```
*** DETAILS: Launching the job. ***********************************************
POST https://192.168.4.14/api/v1/job_templates/272/launch/
Data: {u'job_tags': u'default,npm,pm2,env'}
```
But when asking for the job status, you'll see the tags have gone missing (see "job_tags"):
```
*** DETAILS: Asking for job status. *******************************************
GET https://192.168.4.14/api/v1/jobs/15270/
Params: {}

{
  "id": 15270,
  "type": "job",
  "url": "/api/v1/jobs/15270/",

< snip non-interesting json >

  "created": "2016-04-15T11:13:16.995Z",
  "modified": "2016-04-15T11:13:17.055Z",
  "name": "dev.exonet.nl deploy",
  "description": "",
  "job_type": "run",
  "inventory": 1,
  "project": 88,
  "playbook": "projects_deployment/main.yml",
  "credential": 1,
  "cloud_credential": null,
  "forks": 0,
  "limit": "",
  "verbosity": 3,
  "extra_vars": "{}",
  "job_tags": "",
  "force_handlers": false,
  "skip_tags": "",
  "start_at_task": "",
  "unified_job_template": 272,
  "launch_type": "manual",
  "status": "pending",
  "failed": false,
  "started": null,
  "finished": null,
  "elapsed": 0.0,
  "job_args": "",
  "job_cwd": "",
  "job_env": {},
  "job_explanation": "",
  "result_stdout": "Waiting for results...",
  "result_traceback": "",
  "job_template": 272,
  "passwords_needed_to_start": [],
  "ask_variables_on_launch": false,
  "changed": true
}
```
If I look at the REST api, there's no such thing as tags for the launch command of a job_template:

```
OPTIONS /api/v1/job_templates/272/launch/
HTTP 200 OK
Content-Type: application/json
Vary: Accept
Allow: GET, POST, HEAD, OPTIONS
X-API-Time: 0.093s


{
    "name": "Job Template Launch", 
    "description": "Launch a Job Template:\n\nMake a GET request to this resource to determine if the job_template can be\nlaunched and whether any passwords are required to launch the job_template.\nThe response will include the following fields:\n\n* `ask_variables_on_launch`: Flag indicating whether the job_template is\n  configured to prompt for variables upon launch (boolean, read-only)\n* `can_start_without_user_input`: Flag indicating if the job_template can be\n  launched without user-input (boolean, read-only)\n* `passwords_needed_to_start`: Password names required to launch the\n  job_template (array, read-only)\n* `variables_needed_to_start`: Required variable names required to launch the\n  job_template (array, read-only)\n* `survey_enabled`: Flag indicating if whether the job_template has an enabled\n  survey (boolean, read-only)\n* `credential_needed_to_start`: Flag indicating the presence of a credential\n  associated with the job template.  If not then one should be supplied when\n  launching the job (boolean, read-only)\n\nMake a POST request to this resource to launch the job_template. If any\npasswords or extra variables (extra_vars) are required, they must be passed\nvia POST data, with extra_vars given as a YAML or JSON string and escaped\nparentheses. If `credential_needed_to_start` is `True` then the `credential`\nfield is required as well.\n\nIf successful, the response status code will be 202.  If any required passwords\nare not provided, a 400 status code will be returned.  If the job cannot be\nlaunched, a 405 status code will be returned.", 
    "renders": [
        "application/json", 
        "text/html"
    ], 
    "parses": [
        "application/json"
    ], 
    "added_in_version": "1.2", 
    "actions": {
        "POST": {
            "extra_vars": {
                "type": "string", 
                "required": false, 
                "label": "extra vars", 
                "default": "", 
                "write_only": true
            }, 
            "credential": {
                "type": "field", 
                "required": false, 
                "default": null, 
                "write_only": true
            }
        }, 
```
There is only `extra_vars`. 

I think passing --tags onto tower-cli should result in a POST to /api/v1/jobs/. There we can specify tags and extra_vars and base the job on a template:

```
{
    "name": "", 
    "description": "", 
    "job_type": "run", 
    "inventory": null, 
    "project": null, 
    "playbook": "", 
    "credential": null, 
    "cloud_credential": null, 
    "forks": 0, 
    "limit": "", 
    "verbosity": 0, 
    "extra_vars": "", 
    "job_tags": "", 
    "force_handlers": false, 
    "skip_tags": "", 
    "start_at_task": "", 
    "job_template": null
}
```

**Comment by AlanCoding** [comment]
Great point. Thanks for bringing this up. I agree the `job_tags` option shouldn't be present for this command in its present form.

I have conflicted feelings about switching the POST to the /jobs/ endpoint. That seems more like a separate command to me because it is inconsistent with using the `--job-template` flag. But maybe it would be appropriate inside of the `job launch` command if job template was not provided, or if some other flag was given to specify use of the /jobs/ endpoint.

**Comment by styxit** [comment]
I did some research on this. Apparently the `tower-cli job launch` command checks if the tower version support launching job_templates. If so, the job_template is launched. If not a new job is created and after that, the job is started. That way passing tags works fine.

In my test i forced the job launch command to never use the `/job_templates/%d/launch` endpoint, but do it the pre tower2.1 way. Creating a job based on a template and after that starting it. That way passing tags to tower works fine.

I would recommend to not use the `/job_templates/%d/launch` endpoint in tower-cli because the only options it supports are `extra_vars`and `credential`. While the `POST jobs` and `GET /jobs/%d/start` support all kinds of parameters for starting jobs.

The problem with this is, since tower 2.1 non-admin users can not create jobs, only launch templates.
Tower 2.1 changelog:
> Non-admin api users must now use the /launch endpoint for a job template and can no longer call a job’s /start endpoint directly.

It was unclear to me what a `Non-admin api user` is. But i tried different things and i guess this is actually a Tower super-user and not just an organisation admin. Because accessing `POST /jobs` was not allowed for an my organisation-admin user. I am not 100% sure about this though.

In short:
- If the template /launch endpoint exists, a non-admin user *must* use it to start a job since tower 2.1. Job tags can not be provided.
- An admin user prefers to use the job /start endpoint to be able to pass job_tags.
- If the template /launch endpoint does not exist, use job /start for all users.

To achieve this, a check to `/me` can be made to verify if the user has superadmin privileges. Unfortunately this would introduce an extra api call.

My solution: ansible/tower-cli#/137

**Comment by jangsutsr** [comment]
@rvdh We added a new flag option `--use-job-endpoint` for `job_template launch` command which will come with tower-cli 3.0.0. You can use this flag to manually disable `/api/v1/job_template/` endpoint.
